### PR TITLE
Trial post processing

### PR DESCRIPTION
**Pull request by hvy**
## Motivation

~Depends on https://github.com/optuna/optuna/pull/2133.~ (edit: merged)

Allow experimental trial post post-processing. As a consequence, the `BoTorchSampler` constraints callback assumptions are relaxed and they no longer have to be deterministic since they'll only be computed once per trial. The `GridSampler` can also make use of this change to stop the study.

## Description of the changes

Introduces `BaseSampler.after_trial` to allow trial post-processing by each sampler.

## TODO

- [x] Make sure API looks okay. Ensure compatibility with possible future ask-and-tell API.
- [x] Tests. Verify error handling and edge cases.

### CODE REVIEW

File: optuna/samplers/_base.py [hunk 1]
```diff
@@ -137,6 +140,37 @@ def sample_independent(
 
         raise NotImplementedError
 
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Optional[Sequence[float]],
```
Comment (hvy) [inline optuna/samplers/_base.py:148]: Just accepting a `FrozenTrial` with its state and values set would speak for a cleaner API. This is however not possible since to the storage cache mechanism assumes that a finished trial is never modified.

**Comment by hvy** [comment]
Exceptions raised from `after_trial` are not affected by the `catch` argument to `Study.optimize` and will always be propagated to the the optimize caller. This makes sense if we for instance would want to introduce a "before_trial" to e.g. check for SO/MO compatibility raising appropriate errors. These errors should likely not be affected by `catch` neither.

**Comment by hvy** [comment]
Removed WIP, PTAL.

File: optuna/samplers/_base.py [hunk 2]
```diff
@@ -137,6 +140,37 @@ def sample_independent(
...
+    ) -> None:
+        """Trial post-processing.
+
+        This method is called after the objective function returns and right before the trials is
+        finished and its state is stored.
+
+        .. note::
+            Added in v2.4.0 as an experimental feature. The interface may change in newer versions
+            without prior notice. See https://github.com/optuna/optuna/releases/tag/v2.4.0.
```
Comment (hvy) [inline optuna/samplers/_base.py:157]: Avoided using the `@experimental` decorator since this method is called internally and we don't want actual warnings being propagated to the user. We can silence warning internally upon invocations but given that particular samplers could raise actual warnings, might not be very smart.

File: optuna/samplers/_grid.py [hunk 3]
```diff
@@ -185,6 +179,23 @@ def sample_independent(
...
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        target_grids = self._get_unvisited_grid_ids(study)
+
+        if len(target_grids) == 0:
+            study.stop()
+        elif len(target_grids) == 1:
+            if state == TrialState.COMPLETE:
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/samplers/_grid.py:194]: Thanks, makes sense. I also added a test for this case in the latest commit.

File: optuna/integration/botorch.py [hunk 4]
```diff
@@ -488,8 +459,33 @@ def sample_relative(
...
+            if con is None:
+                warnings.warn(
+                    "`constraints_func` was given but no call to it correctly computed "
+                    "constraints. Constraints passed to `candidates_func` will be `None`."
+                )
+            elif numpy.isnan(con).any():
+                warnings.warn(
+                    "`constraints_func` was given but some calls to it did not correctly compute "
+                    "constraints. Constraints passed to `candidates_func` will contain NaN."
+                )
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:488]: It seems that there is no test for this case. How about adding it?

File: optuna/integration/botorch.py [hunk 5]
```diff
@@ -555,3 +551,31 @@ def sample_independent(
...
+
+            try:
+                con = self._constraints_func(trial)
+                if not isinstance(con, (tuple, list)):
+                    warnings.warn(
+                        f"Constraints should be a sequence of floats but got {constraints}."
+                    )
+                constraints = tuple(con)
+            except Exception:
+                raise
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:573]: If calculated `con` cannot be cast to `tuple`, the error will be raised. How about testing it?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. Basically, LGTM. I have some early comments on tests.

File: optuna/samplers/_grid.py [hunk 6]
```diff
@@ -185,6 +179,23 @@ def sample_independent(
...
+    ) -> None:
+        target_grids = self._get_unvisited_grid_ids(study)
+
+        if len(target_grids) == 0:
+            study.stop()
+        elif len(target_grids) == 1:
+            if state.is_finished():
+                grid_id = study._storage.get_trial_system_attrs(trial._trial_id)["grid_id"]
+                if grid_id == target_grids[0]:
+                    study.stop()
```
Comment (g-votte) [inline optuna/samplers/_grid.py:197]:
```suggestion
            grid_id = study._storage.get_trial_system_attrs(trial._trial_id)["grid_id"]
            if grid_id == target_grids[0]:
                study.stop()
```

Thank your for applying my previous suggestion.

I've noticed that `if state.is_finished():` is now redundant because the state should be either of `COMPLETE`, `PRUNED`, or `FAIL` as long as `after_trial()` is triggered.

Sorry for suggesting for the same point twice.

**Review by g-votte — CHANGES_REQUESTED** [review]

File: optuna/samplers/_grid.py (hunk 6 above)
Comment (hvy) [reply optuna/samplers/_grid.py:197]: You're right. Thanks.

File: optuna/integration/botorch.py (hunk 5 above)
Comment (hvy) [reply optuna/integration/botorch.py:573]: Ah, it should be tested already by `test_botorch_constraints_func_invalid_type`.

File: optuna/integration/botorch.py (hunk 4 above)
Comment (hvy) [reply optuna/integration/botorch.py:488]: Sure, let me add tests to check the elements for NaN as well.

Comment (hvy) [reply optuna/integration/botorch.py:488]: Added a test for the warning.

**Comment by hvy** [comment]
Thanks for the review comments, PTAL.

File: optuna/_optimize.py [hunk 7]
```diff
@@ -219,26 +219,31 @@ def _run_trial(
...
-                    trial.number, repr(func_err)
-                ),
-                exc_info=func_err_fail_exc_info,
-            )
-        elif values_conversion_failure_message is not None:
-            _logger.warning(values_conversion_failure_message)
+    try:
+        trial._after_func(state, values)
+    except Exception:
+        raise
```
Comment (HideakiImamura) [inline optuna/_optimize.py:225]: It is more consistent to chain the exception like https://github.com/optuna/optuna/pull/1781.

Comment (hvy) [reply optuna/_optimize.py:225]: Hm, #1781 fixes exceptions to not silence caught exceptions. Here, we're propagating and not silencing anything. Do you mean to construct an error here and chain with it? I think there are pros and cons but ss can be seen in #1781, we don't do that too often. In my understanding, it'll be more to document and understand from the user perspective.

Comment (HideakiImamura) [reply optuna/_optimize.py:225]: Thanks for the detailed explanation. I understood. We should not construct an error here.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the discussions. LGTM.

**Review by toshihikoyanase — COMMENTED** [review]
It looks great to me. I add some small comments, but I think they are optional.

**Review by g-votte — APPROVED** [review]
I mainly reviewed the logic of the grid sampler, and it looks good to me.

@toshihikoyanase  
Please merge after your review.

File: optuna/integration/botorch.py [hunk 8]
```diff
@@ -488,8 +459,33 @@ def sample_relative(
...
+                        )
+
+                    con[trial_idx] = constraints
+
+        if self._constraints_func is not None:
+            if con is None:
+                warnings.warn(
+                    "`constraints_func` was given but no call to it correctly computed "
+                    "constraints. Constraints passed to `candidates_func` will be `None`."
+                )
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/integration/botorch.py:483]: Added a test.

File: optuna/integration/botorch.py [hunk 9]
```diff
@@ -488,8 +459,33 @@ def sample_relative(
                     value *= -1
                 values[trial_idx, obj_idx] = value
 
-            if con is not None:
-                con[trial_idx] = self._trial_constraints[trial_idx]
+            if self._constraints_func is not None:
+                constraints = study._storage.get_trial_system_attrs(trial._trial_id)[
+                    "botorch:constraints"
+                ]
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/integration/botorch.py]: Thanks, fixed the logic and added a test.

**Comment by hvy** [comment]
Thanks for the review comments again, they should be addressed. PTAL.

The tests are quite involved now for the `BoTorchSampler` even though it really wasn't the main focus of this PR. Given that it's an experimental sampler, please don't hesitate to make changes to those tests as needed.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. LGTM!
